contracts/utils/KlimaCarbonRetirements.sol

```diff
@@ -0,0 +1,287 @@
+        );
+        _transferOwnership(newOwner);
+    }
+
+    /**
+     * @dev Transfers ownership of the contract to a new account (`newOwner`).
+     * Internal function without access restriction.
+     */
+    function _transferOwnership(address newOwner) internal virtual {
+        address oldOwner = _owner;
+        _owner = newOwner;
+        emit OwnershipTransferred(oldOwner, newOwner);
+    }
+}
+
+// File: contracts/utils/Ownable.sol
+
+// File: contracts/KlimaCarbonRetirements.sol
+
+// SPDX-License-Identifier: MIT
+
+pragma solidity ^0.8.10;
+
+/**
+    @title Klima Retirement Storage
+    @notice This is used to store any offset retirements made through Klima retirement helper contracts.
+ */
+contract KlimaCarbonRetirements is Ownable {
+    struct Retirement {
+        uint256 totalRetirements;
+        uint256 totalCarbonRetired;
+        uint256 totalClaimedNFT;
+        mapping(uint256 => address) retiredPool;
+        mapping(uint256 => uint256) retiredAmount;
+        mapping(uint256 => string) retirementBeneficiary;
+        mapping(address => uint256) totalPoolRetired;
+    }
+
+    mapping(address => Retirement) public retirements;
+    mapping(address => bool) public isHelperContract;
+    mapping(address => bool) public isMinterContract;
+
+    uint256 totalCarbonRetired;
+
+    /**
+        @notice Stores the details of an offset transaction for future use
+        @param _retiree Address of the retiree. Not the address of a helper contract.
+        @param _pool Address of the carbon pool token.
+        @param _amount Number of tons offset. Expected is with 18 decimals.
+     */
+    function carbonRetired(
+        address _retiree,
+        address _pool,
+        uint256 _amount
+    ) public returns (bool) {
+        require(
+            isHelperContract[msg.sender],
+            "Caller is not a defined helper contract"
+        );
+
+        Retirement storage info = retirements[_retiree];
+
+        info.retiredPool[info.totalRetirements] = _pool;
+        info.retiredAmount[info.totalRetirements] = _amount;
+        info.totalPoolRetired[_pool] += _amount;
+        info.totalCarbonRetired += _amount;
+        info.totalRetirements += 1;
+    }
+
+    /**
+        @notice Return any unclaimed NFT totals for an address
+        @param _minter Address of user trying to mint.
+        @return The net amount of offsets not used for minting an NFT to date.
+     */
+    function getUnclaimedTotal(address _minter) public view returns (uint256) {
+        return
+            retirements[_minter].totalCarbonRetired -
+            retirements[_minter].totalClaimedNFT;
+    }
+
+    /**
+        @notice This function updates the total claimed amount for minting an NFT.
+        @param _minter Address of the user trying to mint.
+        @param _amount Amount being claimed for the mint. Expected value in 18 decimals.
+     */
+    function offsetClaimed(address _minter, uint256 _amount)
```

Comment:
Shouldn't this function be gated properly? Seems like anyone can call it? Also shouldn't it be executed inside the same context as NFT minting?

Reply:
```
require(
            isMinterContract[_minter],
            "Calling contract is not an approved minter."
        );
```
Not sure if this has been added afterwards.

Reply:
Yeap, it's there, nevermind.